src/Compilers/CSharp/Portable/Lowering/AsyncRewriter/AsyncRewriter.AsyncIteratorRewriter.cs

```diff
@@ -0,0 +1,557 @@
+
+                // local._valueOrEndPromise = new ManualResetValueTaskSourceLogic<bool>(stateMachine);
+                MethodSymbol mrvtslCtor =
+                    F.WellKnownMethod(WellKnownMember.System_Threading_Tasks_ManualResetValueTaskSourceLogic_T__ctor)
+                    .AsMember((NamedTypeSymbol)_promiseOfValueOrEndField.Type);
+
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), _promiseOfValueOrEndField.AsMember(frameType)),
+                        F.New(mrvtslCtor, F.Local(stateMachineVariable))));
+
+                // PROTOTYPE(async-streams): Why do we need AsMember?
+                // local._promiseIsActive = true;
+                bodyBuilder.Add(
+                    F.Assignment(
+                        F.Field(F.Local(stateMachineVariable), _promiseIsActiveField.AsMember(frameType)),
+                        F.Literal(true)));
+
+                // return local.$stateField;
+                bodyBuilder.Add(F.Return(F.Local(stateMachineVariable)));
+
+                return F.Block(
+                    ImmutableArray.Create(builderVariable),
+                    bodyBuilder.ToImmutableAndFree());
+            }
+
+            /// <summary>
+            /// Generates the WaitForNextAsync method.
+            /// </summary>
+            private void GenerateIAsyncEnumeratorImplementation_WaitForNextAsync()
+            {
+                // Produce the implementation for `ValueTask<bool> WaitForNextAsync()`:
+                // if (State == StateMachineStates.FinishedStateMachine)
+                // {
+                //     return default(ValueTask<bool>)
+                // }
+                // if (!this._promiseIsActive || this.State == StateMachineStates.NotStartedStateMachine)
+                // {
+                //     var inst = this;
+                //     this._builder.Start(ref inst);
+                // }
+                // return new ValueTask<bool>(this, _valueOrEndPromise.Version);
+
+                NamedTypeSymbol IAsyncEnumeratorOfElementType =
+                    F.WellKnownType(WellKnownType.System_Collections_Generic_IAsyncEnumerator_T)
+                    .Construct(_currentField.Type);
+
+                MethodSymbol IAsyncEnumerableOfElementType_WaitForNextAsync =
+                    F.WellKnownMethod(WellKnownMember.System_Collections_Generic_IAsyncEnumerator_T__WaitForNextAsync)
+                    .AsMember(IAsyncEnumeratorOfElementType);
+
+                // The implementation doesn't depend on the method body of the iterator method.
+                OpenMethodImplementation( IAsyncEnumerableOfElementType_WaitForNextAsync, hasMethodBodyDependency: false);
+
+                BoundFieldAccess promiseIsActiveField = F.Field(F.This(), _promiseIsActiveField);
+                LocalSymbol instSymbol = F.SynthesizedLocal(this.stateMachineType);
```

Comment:
>instSymbol [](start = 28, length = 10)

Does `inst` need to be declared in the outermost block? If not consider hiding it completely in `GenerateCallStart()`.